src/java.base/unix/native/criuengine/criuengine.cpp

```diff
@@ -63,6 +66,52 @@ static int kickjvm(pid_t jvm, int code) {
     return 0;
 }
 
+static std::string join_args(const std::vector<const char *> &args) {
+    std::string retval;
+    for (const char *s : args) {
+        if (!s) {
+            continue;
+        }
+        if (!retval.empty()) {
+            retval += ' ';
+        }
+        // https://unix.stackexchange.com/a/357932/296319
+        if (!strpbrk(s, " \t\n!\"#$&'()*,;<=>?[\\]^`{|}~")) {
+            retval += s;
+            continue;
+        }
+        retval += '\'';
+        for (; *s; ++s) {
+            if (*s != '\'') {
+                retval += *s;
+            } else {
+                retval += "'\\''";
+            }
+        }
+        retval += '\'';
+    }
+    return retval;
+}
+
+static std::string path_from(std::string cwd, std::string path) {
+    return path[0] == '/' ? path : cwd + "/" + path;
+}
+
+static std::string path_abs(std::string rel) {
+    char *cwd_s = get_current_dir_name();
+    if (!cwd_s) {
+        perror("get_current_dir_name");
+        exit(1);
+    }
+    std::string cwd = cwd_s;
+    free(cwd_s);
+    return path_from(cwd, rel);
+}
```

Comment:
The problem with this function as abstraction is that it first calculates CWD, then checks was the argument absolute or relative.

Reply:
The code was more simple that way and it was straightforward to replace it with C++17 `std::filesystem::path` after [OpenJDK makes the switch to C++17](https://bugs.openjdk.org/browse/JDK-8310260).

Reply:
The interface is fine, the implementation just could be more straightforward and efficient.

The link you're refering has Resolution: Rejected.